### PR TITLE
Job to Remove Validation Reports for Unsupported Resource Types

### PR DESCRIPTION
**Pull request by JVickery-TBS**
feat(dev): implemented IDomainObjectModification;

- Implemented `IDomainObjectModification`.
- Queued new job callback for unsupported validation resources.
- Added new config option `ckanext.validation.limit_validation_reports`.

See https://github.com/open-data/ckanext-canada/pull/1382 for more discussion.

### CODE REVIEW

**Comment by JVickery-TBS** [comment]
Just working on some better conditions to prevent multiple jobs queuing up

**Comment by JVickery-TBS** [comment]
@wardi okay this should be good for review.

So for the validation extension, I was not able to do it in the `IDomainObjectModification` implementation as it was firing off during the validation runs. And as the validation runs do a resource/package patch it would add another job to the queue, doubling the job.

So I was able to just add it into the `after_update` hook as this extension has a context key in there which handles the resource patch thing.

**Comment by JVickery-TBS** [comment]
@wardi I was also able to fix the KeyErrors happening frequently in the `resource_validation_delete` action method in the `logic.py` script. So the validation extras should actually get deleted from the resource metadata. Hopefully this issue was not happening a lot on production.

File: ckanext/validation/logic.py [hunk 1]
```diff
@@ -217,12 +217,16 @@ def resource_validation_delete(context, data_dict):
...
+            resource.pop(u'validation_status')
+        if resource.get('validation_options', False):
+            resource.pop(u'validation_options')
+        if resource.get('validation_timestamp', False):
+            resource.pop(u'validation_timestamp')
+        user = t.get_action('get_site_user')({'ignore_auth': True})
         t.get_action('resource_update')(
             {'ignore_auth': True,
-             'user': t.get_action('get_site_user')({'ignore_auth': True})['name']},
+             'user': user.get('name', 'ckan-system')},
```
Comment (wardi) [inline ckanext/validation/logic.py]: a user will always have a name

File: ckanext/validation/logic.py [hunk 2]
```diff
@@ -217,12 +217,16 @@ def resource_validation_delete(context, data_dict):
...
     try:
-        resource.pop(u'validation_status')
-        resource.pop(u'validation_options')
-        resource.pop(u'validation_timestamp')
+        if resource.get('validation_status', False):
+            resource.pop(u'validation_status')
+        if resource.get('validation_options', False):
+            resource.pop(u'validation_options')
+        if resource.get('validation_timestamp', False):
+            resource.pop(u'validation_timestamp')
```
Comment (wardi) [inline ckanext/validation/logic.py]: Is there a reason you want to keep only Falsey-valued validation fields here?

File: ckanext/validation/logic.py (hunk 1 above)
Comment (wardi) [reply ckanext/validation/logic.py]: If you're already adding ckanapi elsewhere you could replace all this with a default LocalCKAN and `action.resource_update(resource)`

File: ckanext/validation/plugin.py [hunk 3]
```diff
@@ -316,3 +321,43 @@ def _run_async_validation(resource_id):
         log.warning(
             u'Could not run validation for resource {}: {}'.format(
                 resource_id, str(e)))
+
+
+def _should_remove_unsupported_resource_validation_reports(res_dict):
+    if not t.h.asbool(t.config.get('ckanext.validation.limit_validation_reports', False)):
```
Comment (wardi) [inline ckanext/validation/plugin.py]: Remember to document

```suggestion
    if not t.h.asbool(t.config.get('ckanext.validation.clean_validation_reports', False)):
```

File: ckanext/validation/logic.py (hunk 2 above)
Comment (JVickery-TBS) [reply ckanext/validation/logic.py]: Hmmm no. Wanted to say `if in resource` as the pop was throwing key errors. Which was weird to me, but I do not use pop so not sure if it checks if the key exists in the dict or not before hand.

Will fix.

Comment (JVickery-TBS) [reply ckanext/validation/logic.py]: Okay this should be good now. Things have been moved outside the try/catch and the action now uses `ckanapi.localCKAN`

File: ckanext/validation/logic.py (hunk 1 above)
Comment (JVickery-TBS) [reply ckanext/validation/logic.py]: Accomplished.

**Comment by JVickery-TBS** [comment]
@wardi New config option has been added to readme. And I fixed up the custom code in the `resource_validation_delete` method a bit more. So the key checks are actual key checks now for the validation metadata. And the action now uses the `ckanapi.localCKAN`

**Review by wardi — APPROVED** [review]

**Pull request merged**